File: WorkerNode/src/main/java/com/yahoo/sdvornik/sorter/QuickSortByteArr.java
package com.yahoo.sdvornik.sorter;

import io.netty.buffer.ByteBuf;

import java.nio.ByteBuffer;
import java.util.Arrays;

/**
 * Implementation of QuickSort algorithm for long keys in byte array.
 * For testing purposes.
 */
public class QuickSortByteArr {

    private final byte[] byteArr;

    private final ByteBuffer longBuffer = ByteBuffer.allocate(Long.BYTES);

    public QuickSortByteArr(ByteBuf byteBuf) {
        this.byteArr = byteBuf.array();
    }

    public void sortByteBuf() {
        quickSort(0 , byteArr.length/Long.BYTES-1);
    }

    /**
     * QuickSort recursive method for ByteBuf.
     * @param leftBorderIndex
     * @param rightBorderIndex
     */
    private void quickSort(int leftBorderIndex , int rightBorderIndex) {
        if (leftBorderIndex >= rightBorderIndex) return;
        int pivotElmIndex = getPivotElmIndex(leftBorderIndex, rightBorderIndex);
        quickSort(leftBorderIndex, pivotElmIndex - 1);
        quickSort(pivotElmIndex + 1, rightBorderIndex);
    }

    /**
     * Return new pivot element index in [startIndex, endIndex] for long[].
     * @param startIndex
     * @param endIndex
     * @return
     */
    private int getPivotElmIndex(int startIndex, int endIndex ){
        byte[] longBytes = Arrays.copyOfRange(byteArr, endIndex*Long.BYTES, (1+endIndex)*Long.BYTES);

        longBuffer.put(longBytes);
        longBuffer.flip();
        long pivotElmValue = longBuffer.getLong();
        longBuffer.clear();
        int pivotElmIndex = startIndex;
        for(int i = startIndex; i<endIndex; ++i){
            longBytes = Arrays.copyOfRange(byteArr, i*Long.BYTES, (1+i)*Long.BYTES);
            longBuffer.put(longBytes);
            longBuffer.flip();
            long elm = longBuffer.getLong();
            longBuffer.clear();
            if (elm >= pivotElmValue) continue;
            swapElmInByteArr(pivotElmIndex, i);
            ++pivotElmIndex;
        }
        swapElmInByteArr(pivotElmIndex,endIndex);
        return pivotElmIndex;
    }

    protected void swapElmInByteArr(int i, int j){
        byte[] tempI = Arrays.copyOfRange(byteArr, i*Long.BYTES, (i+1)*Long.BYTES);
        byte[] tempJ = Arrays.copyOfRange(byteArr, j*Long.BYTES, (j+1)*Long.BYTES);
        for(int k = 0; k < Long.BYTES; ++k) {
            byteArr[i*Long.BYTES+k] = tempJ[k];
            byteArr[j*Long.BYTES+k] = tempI[k];
        }
    }
}

File: worker/src/main/kotlin/com/yahoo/sdvornik/sorter/QuickSortByteArr.kt
package com.yahoo.sdvornik.sorter

import io.netty.buffer.ByteBuf
import java.nio.ByteBuffer
import java.util.*

/**
 * @author Serg Dvornik <sdvornik@yahoo.com>
 */
class QuickSortByteArr(byteBuf: ByteBuf) {

    private val byteArr: ByteArray

    private val longBuffer = ByteBuffer.allocate(java.lang.Long.BYTES)

    init {
        this.byteArr = byteBuf.array()
    }

    fun sortByteBuf() {
        quickSort(0, byteArr.size / java.lang.Long.BYTES - 1)
    }

    /**
     * QuickSort recursive method for ByteBuf.
     * @param leftBorderIndex
     * @param rightBorderIndex
     */
    private fun quickSort(leftBorderIndex: Int, rightBorderIndex: Int) {
        if (leftBorderIndex >= rightBorderIndex) return
        val pivotElmIndex = getPivotElmIndex(leftBorderIndex, rightBorderIndex)
        quickSort(leftBorderIndex, pivotElmIndex - 1)
        quickSort(pivotElmIndex + 1, rightBorderIndex)
    }

    /**
     * Return new pivot element index in [startIndex, endIndex] for long[].
     * @param startIndex
     * @param endIndex
     * @return
     */
    private fun getPivotElmIndex(startIndex: Int, endIndex: Int): Int {
        var longBytes = Arrays.copyOfRange(byteArr, endIndex * java.lang.Long.BYTES, (1 + endIndex) * java.lang.Long.BYTES)

        longBuffer.put(longBytes)
        longBuffer.flip()
        val pivotElmValue = longBuffer.long
        longBuffer.clear()
        var pivotElmIndex = startIndex
        for (i in startIndex..endIndex - 1) {
            longBytes = Arrays.copyOfRange(byteArr, i * java.lang.Long.BYTES, (1 + i) * java.lang.Long.BYTES)
            longBuffer.put(longBytes)
            longBuffer.flip()
            val elm = longBuffer.long
            longBuffer.clear()
            if (elm >= pivotElmValue) continue
            swapElmInByteArr(pivotElmIndex, i)
            ++pivotElmIndex
        }
        swapElmInByteArr(pivotElmIndex, endIndex)
        return pivotElmIndex
    }

    protected fun swapElmInByteArr(i: Int, j: Int) {
        val tempI = Arrays.copyOfRange(byteArr, i * java.lang.Long.BYTES, (i + 1) * java.lang.Long.BYTES)
        val tempJ = Arrays.copyOfRange(byteArr, j * java.lang.Long.BYTES, (j + 1) * java.lang.Long.BYTES)
        for (k in 0..java.lang.Long.BYTES - 1) {
            byteArr[i * java.lang.Long.BYTES + k] = tempJ[k]
            byteArr[j * java.lang.Long.BYTES + k] = tempI[k]
        }
    }
}
